crates/aptos-crypto/src/p256_ecdsa/p256_keys.rs

```diff
@@ -0,0 +1,275 @@
+            Ok(p256_secret_key) => Ok(P256PrivateKey(p256_secret_key)),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+
+    /// Private function aimed at minimizing code duplication between sign
+    /// methods of the SigningKey implementation. This should remain private.
+    /// This function uses the `RustCrypto` P256 signing library, which uses,
+    /// as of version 0.13.2, SHA2-256 as its hashing algorithm
+    fn sign_arbitrary_message(&self, message: &[u8]) -> P256Signature {
+        let secret_key = &self.0;
+        let sig = P256Signature(secret_key.sign(message.as_ref()));
+        P256Signature::make_canonical(&sig)
+    }
+}
+
+impl P256PublicKey {
+    /// Serialize a P256PublicKey. Uses the SEC1 serialization format.
+    pub fn to_bytes(&self) -> [u8; P256_PUBLIC_KEY_LENGTH] {
+        // The RustCrypto P256 `to_sec1_bytes` call here should never return an array of the wrong length and cause a panic
+        (*self.0.to_sec1_bytes()).try_into().unwrap()
+    }
+
+    /// Deserialize a P256PublicKey, checking expected key size
+    /// and that it is a valid curve point.
+    /// Uses the SEC1 serialization format.
+    pub(crate) fn from_bytes_unchecked(
+        bytes: &[u8],
+    ) -> std::result::Result<P256PublicKey, CryptoMaterialError> {
+        match p256::ecdsa::VerifyingKey::from_sec1_bytes(bytes) {
+            Ok(p256_public_key) => Ok(P256PublicKey(p256_public_key)),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+}
+
+///////////////////////
+// PrivateKey Traits //
+///////////////////////
+
+impl PrivateKey for P256PrivateKey {
+    type PublicKeyMaterial = P256PublicKey;
+}
+
+impl SigningKey for P256PrivateKey {
+    type SignatureMaterial = P256Signature;
+    type VerifyingKeyMaterial = P256PublicKey;
+
+    fn sign<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+    ) -> Result<P256Signature, CryptoMaterialError> {
+        Ok(P256PrivateKey::sign_arbitrary_message(
+            self,
+            signing_message(message)?.as_ref(),
+        ))
+    }
+
+    #[cfg(any(test, feature = "fuzzing"))]
+    fn sign_arbitrary_message(&self, message: &[u8]) -> P256Signature {
+        P256PrivateKey::sign_arbitrary_message(self, message)
+    }
+}
+
+impl Uniform for P256PrivateKey {
+    // Returns a random field element as a private key indistinguishable from uniformly random.
+    // Uses a hack to get around the incompatability of the `aptos-crypto` RngCore trait and the
+    // `RustCrypto` RngCore trait
+    fn generate<R>(rng: &mut R) -> Self
+    where
+        R: ::rand::RngCore + ::rand::CryptoRng + ::rand_core::CryptoRng + ::rand_core::RngCore,
+    {
+        let mut bytes = [0u8; P256_PRIVATE_KEY_LENGTH * 2];
+        rng.fill_bytes(&mut bytes);
+        let bignum = BigUint::from_bytes_be(&bytes[..]);
+        let order = BigUint::from_bytes_be(&ORDER);
```

Comment:
I hope the LE/BE encoding of `ORDER` here is right. If we get these flipped, the key space from which we sample could be much smaller and disaster will ensue.

Are there any tests we can do here?

Reply:
Yeah I'm trying to figure this out. The encoding for `RustCrypto` is surprisingly opaque

Reply:
Ok converting the array `[0,0,0,1]` to a `BigUint` with `from_bytes_be` then printing the result gives `[1]`, while converting the same array and printing the resulting `BigUint` with `from_bytes_le` gives `[16777216]`. So `_be` is correct here since the `ORDER` constant array is also has the more important integers on the left of the less important ones

Reply:
Indeed, [`ORDER` is in BE](https://github.com/RustCrypto/elliptic-curves/blob/60664a93bc17c866974f145c04905d2bab87e555/p256/src/lib.rs#L108).